File: leetcode/java/54.spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        if(matrix == null || matrix.length == 0 || matrix[0] == null || matrix[0].length == 0) return new ArrayList<>(); 
        List<Integer> res = new ArrayList<>();
        int rowh =  matrix.length - 1;
        int colr =  matrix[0].length - 1;
        int rowl = 0;
        int coll = 0;
        while(rowl <= rowh && coll <= colr){
            for(int i = coll; i <= colr; i++){
                res.add(matrix[rowl][i]);
            }
            rowl++;
            for(int i = rowl; i <= rowh; i++){
                res.add(matrix[i][colr]);
            }
            colr--;
            if(rowl <= rowh){
                
            for(int i = colr; i >=  coll; i--){
                res.add(matrix[rowh][i]);
            }
            rowh--;
            }
            if(coll <= colr){
                for(int i = rowh; i >= rowl; i--){
                    res.add(matrix[i][coll]);
                }
                coll++;
                
            }
        }
        return res;
        
            
    }
}

File: leetcode/golang/54.spiral-matrix.go
package main

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return []int{}
	}
	m := len(matrix)
	n := len(matrix[0])
	res := []int{}
	up := 0
	down := m - 1
	left := 0
	right := n - 1
	for up <= down && left <= right {
		for j := left; j <= right; j++ {
			res = append(res, matrix[up][j])
		}
		up++

		for i := up; i <= down; i++ {
			res = append(res, matrix[i][right])
		}
		right--

		if up <= down {
			for j := right; j >= left; j-- {
				res = append(res, matrix[down][j])
			}
		}
		down--

		if left <= right {
			for i := down; i >= up; i-- {
				res = append(res, matrix[i][left])
			}
		}
		left++
	}
	return res
}
